include/nbl/builtin/hlsl/subgroup2/arithmetic_portability_impl.hlsl

```diff
@@ -103,74 +106,98 @@ struct reduction
 
-    type_t operator()(NBL_CONST_REF_ARG(type_t) value)
+    // affected by https://github.com/microsoft/DirectXShaderCompiler/issues/7006
+    // NBL_CONSTEXPR_STATIC_INLINE uint32_t SubgroupSizeLog2 = config_t::SizeLog2;
+
+    type_t operator()(type_t value)
     {
-        op_t op;
-        return op(value);
+        return __call(value);
+    }
+
+    static type_t __call(type_t value)
+    {
+        binop_t op;
+        const uint32_t subgroupInvocation = glsl::gl_SubgroupInvocationID();
+    
+        type_t rhs = glsl::subgroupShuffleUp<type_t>(value, 1u); // all invocations must execute the shuffle, even if we don't apply the op() to all of them
+        // TODO waiting on mix intrinsic fix from bxdf branch, value = op(value, hlsl::mix(rhs, binop_t::identity, subgroupInvocation < 1u));
+        value = op(value, subgroupInvocation<1u ? binop_t::identity : rhs);
+    
+        const uint32_t SubgroupSizeLog2 = config_t::SizeLog2;
+        [unroll]
+        for (uint32_t i = 1; i < integral_constant<uint32_t,SubgroupSizeLog2>::value; i++)
+        {
+            const uint32_t step = i * 2;
+            rhs = glsl::subgroupShuffleUp<type_t>(value, step);
+            // TODO value = op(value, hlsl::mix(rhs, binop_t::identity, subgroupInvocation < step));
+            value = op(value, subgroupInvocation<step ? binop_t::identity : rhs);
+        }
+        return value;
     }
 };
 
-template<class Params, bool native>
-struct exclusive_scan<Params, 1, native>
+template<class Params, class BinOp>
+struct exclusive_scan<Params, BinOp, 1, false>
 {
     using type_t = typename Params::type_t;
     using scalar_t = typename Params::scalar_t;
     using binop_t = typename Params::binop_t;
-    using op_t = subgroup::impl::exclusive_scan<binop_t, native>;
 
     type_t operator()(NBL_CONST_REF_ARG(type_t) value)
     {
-        op_t op;
-        return op(value);
+        value = inclusive_scan<Params, BinOp, 1, false>::__call(value);
+        // can't risk getting short-circuited, need to store to a var
+        type_t left = glsl::subgroupShuffleUp<type_t>(value,1);
+        // the first invocation doesn't have anything in its left so we set to the binop's identity value for exlusive scan
+        return bool(glsl::gl_SubgroupInvocationID()) ? left:binop_t::identity;
     }
 };
 
-template<class Params, bool native>
-struct reduction<Params, 1, native>
+template<class Params, class BinOp>
+struct reduction<Params, BinOp, 1, false>
 {
     using type_t = typename Params::type_t;
     using scalar_t = typename Params::scalar_t;
     using binop_t = typename Params::binop_t;
-    using op_t = subgroup::impl::reduction<binop_t, native>;
 
     scalar_t operator()(NBL_CONST_REF_ARG(type_t) value)
     {
-        op_t op;
-        return op(value);
+        // take the last subgroup invocation's value for the reduction
+        return subgroup::BroadcastLast<type_t>(inclusive_scan<Params, BinOp, 1, false>::__call(value));
```

Comment:
take the code from `inclusive_scan` but remove the ternaries and use `sungroupShuffleXOR` to add items to each other

Basically
```cpp
for (uint32_t i = 0; i < config_t::SizeLog2; i++)
   value = op(glsl::subgroupShuffleXor<type_t>(value,0x1u<<i),value); 
return value;
```